opencog/guile/SchemeSmobAS.cc

```diff
@@ -75,14 +75,16 @@ void SchemeSmob::release_as (AtomSpace *as)
 		deleteable_as.erase(has);
 		lck.unlock();
 		delete as;
+		lck.lock();
 
 		// (Recursively) decrement the use count on the parent.
 		// We had incremented it earlier, in `ss_new_as`, when
-		// creating it.
+		// creating it. Do not delete it, unless the guile gc
+		// asks us to.
 		if (env and deleteable_as.end() != deleteable_as.find(env))
 		{
-			--deleteable_as[env];
-			if (0 == deleteable_as[env]) release_as(env);
+			if (0 < deleteable_as[env])
```

Comment:
Do we really have the situations when `deleteable_as[env] == 0` here? Or it is double check guard? Should we log such situations for analysis?

Reply:
A count of zero means that there are no known C++ users of the thing. There might still be guile references to it, that we don't know of.  The memory can only be collected if there are no known C++ users, and if guile thinks it doesn't have any users.